### DISCUSSION TITLE
[ZK Stack] - Block range limit exceeded

### DISCUSSION THREAD

**Discussion by sachanayush47** [zkSync-Community-Hub/zksync-developers#706] [ZK Stack] [answered]
### Module Affected

- [ ] System contracts (era-system-contracts)
- [ ] Smart contracts (era-contracts)
- [ ] zkEVM (era-zk_evm)
- [X] Server (zksync-era)
- [ ] Prover (era-boojum)
- [ ] Fee withdrawer (era-fee-withdrawer)
- [ ] Block Explorer (clock-explorer)
- [ ] Portal - Wallet + Bridge (dapp-portal)
- [ ] Other (please specify below)

### Rust Version

1.80.1

### Operating System

- [X] Ubuntu
- [ ] MacOS
- [ ] Windows
- [ ] Other (Please specify below)

### Issue Description

Im on chainstack growth plan and it offers only upto 10,000 block size. Is there any way I can cap maximum numbers of blocks that can be queried.

### Expected Behavior

L2 server should be able to run properly.

### Current Behavior

Unable to process requests.

### Additional Details

```
Sep 26 12:21:10 validium-testnet start.sh[55814]: 2024-09-26T12:21:10.624255Z  WARN EthWatch::loop_iteration: zksync_eth_watch::client: Provider returned error message: JSON-RPC request get_logs(filter=Filter { from_block: Some(Number(6712839)), to_block: Some(Number(6763059)), block_hash: None, address: Some(ValueOrArray([0x40fdfd51b7635b1c070596432d4dad117953ef0f, 0x06cb09ced0cb57989356fcdfb3e2ebfe9686eb95, 0xa09a0c677f50164fa9a5d672c1603e38f6e62528, 0x3364b7544cc3648a5ab8b2480576f3d309af1554])), topics: Some([Some(ValueOrArray([0x4531cd5795773d7101c17bdeb9f5ab7f47d7056017506f937083be5d6e77a382]))]), limit: None }) failed: ErrorObject { code: InvalidParams, message: "Block range limit exceeded. See more details at https://docs.chainstack.com/docs/limits#evm-range-limits", data: None }

Sep 26 12:21:10 validium-testnet start.sh[55814]: 2024-09-26T12:21:10.624321Z ERROR zksync_eth_watch: Failed to process new blocks: Eth client error: JSON-RPC request get_logs(filter=Filter { from_block: Some(Number(6712839)), to_block: Some(Number(6763059)), block_hash: None, address: Some(ValueOrArray([0x40fdfd51b7635b1c070596432d4dad117953ef0f, 0x06cb09ced0cb57989356fcdfb3e2ebfe9686eb95, 0xa09a0c677f50164fa9a5d672c1603e38f6e62528, 0x3364b7544cc3648a5ab8b2480576f3d309af1554])), topics: Some([Some(ValueOrArray([0x4531cd5795773d7101c17bdeb9f5ab7f47d7056017506f937083be5d6e77a382]))]), limit: None }) failed: ErrorObject { code: InvalidParams, message: "Block range limit exceeded. See more details at https://docs.chainstack.com/docs/limits#evm-range-limits", data: None }

```

### Prior Research

- [X] Yes, I searched for similar issues and did not find any matching ones.

**Comment by alhenaai** [comment]
It appears you are encountering a block range limit issue due to the constraints set by your infrastructure provider, Chainstack, which limits the block size to 10,000. To address this issue within the ZKsync Era server module, you might consider implementing a mechanism to limit the range of blocks queried at a time to stay within this limit.

Here are a few suggestions:

1. **Modify Query Ranges**: Adjust your queries to request smaller ranges of blocks that do not exceed the 10,000 block limit. This might involve modifying how your server requests data or setting up a more granular polling mechanism.

2. **Batch Processing**: Implement batch processing of blocks in smaller chunks. This can be done by sequentially processing smaller sets of blocks and then aggregating the results as needed.

3. **Optimize Data Requests**: Review and optimize the data requested to ensure that only necessary data is being fetched. This might reduce the frequency and size of the requests.

4. **Infrastructure Adjustment**: If feasible, consider upgrading your plan with Chainstack or switching to a different provider that offers higher limits suitable for your needs.

For more detailed guidance on handling these types of issues, you might want to consult with your development team or reach out to Chainstack support for advice on best practices within their limits.

Your feedback helps us serve you better. If this fully addresses your question, please give it a thumbs up. If it doesn't, please give it a thumbs down. This will inform the community and team so we can improve our support.

**Reply by sachanayush47** [reply]
Where in code I should make changes to fix this?

**Reply by alhenaai** [reply]
To address the block range limit exceeded issue in your ZKsync Era server setup, you'll need to modify the part of your code that handles Ethereum blockchain queries. This typically involves adjusting how logs or blocks are fetched from the Ethereum node. Here’s a general approach on where and how to make these changes:

### 1. Identify the Query Function
Locate the function in your codebase where the Ethereum logs or block data is being fetched. This is likely where you are using functions like `get_logs` or similar Ethereum client calls.

### 2. Implement Pagination or Chunking
Modify the function to fetch data in smaller chunks. For example, if your current code fetches a range of blocks in one go, you can break it down to request smaller ranges sequentially.

**Example Code Adjustment:**
```rust
// Example of fetching logs in smaller chunks
async fn fetch_logs_in_chunks(web3: &Web3, from_block: u64, to_block: u64) -> Result<Vec<Log>, Error> {
    let chunk_size = 10000; // Set based on your provider's limit
    let mut logs = Vec::new();
    let mut start_block = from_block;

    while start_block <= to_block {
        let end_block = std::cmp::min(start_block + chunk_size - 1, to_block);
        let filter = FilterBuilder::default()
            .from_block(BlockNumber::Number(start_block.into()))
            .to_block(BlockNumber::Number(end_block.into()))
            .build();

        let chunk_logs = web3.eth().logs(filter).await?;
        logs.extend(chunk_logs);
        start_block = end_block + 1;
    }

    Ok(logs)
}
```

### 3. Adjust Error Handling
Ensure that your error handling is robust enough to manage partial data or retries in case of failures due to limits being hit.

### 4. Test Changes
After making these changes, thoroughly test the modified code to ensure it handles the block ranges correctly and adheres to the limits set by your infrastructure provider.

### 5. Deploy and Monitor
Deploy the changes to your production environment and monitor the application for any issues related to data fetching. Adjust as necessary based on the operational feedback.

By implementing these changes, you should be able to manage the block range limits more effectively and ensure that your server operates within the constraints set by Chainstack.

Your feedback helps us serve you better. If this fully addresses your question, please give it a thumbs up. If it doesn't, please give it a thumbs down. This will inform the community and team so we can improve our support.

**Reply by popzxc** [reply]
The AI generated answer is not helpful, sorry for that. See the answer below.

**Reply by alhenaai** [reply]
Thank you for your question! I’m unable to provide an answer at this moment, but I have forwarded your inquiry to our team. Someone will be with you shortly to assist further. Your patience is appreciated!

**Comment by popzxc** [comment, marked as answer]
@sachanayush47 Hey. The eth watcher component has internal logic to handle it: if the block range is too big, it would automatically use binary search to scan the range in smaller chunks. [Here's the code for that.](https://github.com/matter-labs/zksync-era/blob/main/core/node/eth_watch/src/client.rs#L148)

However, unfortunately, RPC providers have custom error messages for that, so it needs to be handles on a case by case basis, and right now the chainstack message is not supported.

You can either submit a PR (preferable) or open an issue in the [zksync-era](https://github.com/matter-labs/zksync-era/) repository and ping @perekopskiy.

**Reply by sachanayush47** [reply]
Hi,
Thank you for your help. Problem has been fixed with the help of your solution. I'll surely raise a PR to fix this.

**Reply by sachanayush47** [reply]
> @sachanayush47 Hey. The eth watcher component has internal logic to handle it: if the block range is too big, it would automatically use binary search to scan the range in smaller chunks. [Here's the code for that.](https://github.com/matter-labs/zksync-era/blob/main/core/node/eth_watch/src/client.rs#L148)
> 
> However, unfortunately, RPC providers have custom error messages for that, so it needs to be handles on a case by case basis, and right now the chainstack message is not supported.
> 
> You can either submit a PR (preferable) or open an issue in the [zksync-era](https://github.com/matter-labs/zksync-era/) repository and ping @perekopskiy.

Hi @popzxc , with the help of your solution I was able to fix the issue but I got another issue which I have never faced earlier.
`ERROR zksync_node_framework::service: Task eth_tx_manager panicked: L1 RPC incorrectly reported base prices, either it doesn't return them at all or returns 0's, eth-sender cannot continue without proper base prices!`

I would like change the query that is responsible for querying the L1 sepolia. You can check changes I have made as per your above solution in [PR #2974](https://github.com/matter-labs/zksync-era/pull/2974)

It always starts from the above mentioned number. Can't see the config to change it anywhere.

**Reply by popzxc** [reply]
The error you met [originates in the eth_sender](https://github.com/matter-labs/zksync-era/blob/main/core/node/eth_sender/src/eth_fees_oracle.rs#L39), it means that the `GasAdjuster` provided 0 wei as base fee. It shouldn't normally happen, and may be related to the misconfiguration of the application. If you keep seeing this problem, please open a new discussion and link your comment for context.

Regarding PR -- thank you! I've requested a review from the core team member.

**Reply by perekopskiy** [reply]
Hi @sachanayush47. Regarding "L1 RPC incorrectly reported base prices", it's most probably caused by chainstack not returning blob base fees. We currently don't support such L1 RPC providers.

**Reply by sachanayush47** [reply]
> Hi @sachanayush47. Regarding "L1 RPC incorrectly reported base prices", it's most probably caused by chainstack not returning blob base fees. We currently don't support such L1 RPC providers.

But this issue did not occur when I was using the Chainstack Enterprise plan. At the moment, Im on Growth plan.

**Reply by sachanayush47** [reply]
> Hi @sachanayush47. Regarding "L1 RPC incorrectly reported base prices", it's most probably caused by chainstack not returning blob base fees. We currently don't support such L1 RPC providers.

Can you please let me know which RPC providers are compatible with zksync?

**Reply by perekopskiy** [reply]
APIs that follow [specification](https://ethereum.github.io/execution-apis/api-documentation/)
